File: android/app/src/main/java/io/karpilabs/simplemp3/data/duplicates/DuplicateDetector.kt
package io.karpilabs.simplemp3.data.duplicates

import io.karpilabs.simplemp3.data.local.TrackEntity

/**
 * Groups likely-duplicate rips (same title + artist + duration) so the user can
 * hide extras or merge playlist membership onto a keeper. Never deletes files.
 */
object DuplicateDetector {
    /** Duration bucket width — 2 seconds. */
    const val DURATION_BUCKET_MS: Long = 2_000L

    data class Group(
        val key: String,
        val tracks: List<TrackEntity>,
        val keepId: Long,
    ) {
        val extras: List<TrackEntity> get() = tracks.filter { it.id != keepId }
        val keeper: TrackEntity get() = tracks.first { it.id == keepId }
    }

    fun findGroups(tracks: List<TrackEntity>): List<Group> {
        val songs = tracks.filter { !it.isStream && !it.isHidden }
        val buckets = LinkedHashMap<String, MutableList<TrackEntity>>()
        for (track in songs) {
            val key = fingerprint(track) ?: continue
            buckets.getOrPut(key) { mutableListOf() }.add(track)
        }
        return buckets.values
            .filter { it.size >= 2 }
            .map { group ->
                val ordered = group.sortedWith(keepComparator)
                Group(
                    key = fingerprint(ordered.first()) ?: ordered.first().id.toString(),
                    tracks = ordered,
                    keepId = ordered.first().id,
                )
            }.sortedBy { it.keeper.title.lowercase() }
    }

    fun fingerprint(track: TrackEntity): String? {
        val title = normalize(track.title)
        val artist = normalize(track.artist)
        if (title.isEmpty() || title == "unknown title") return null
        if (artist.isEmpty() || artist == "unknown artist") return null
        val bucket =
            if (track.duration > 0) {
                track.duration / DURATION_BUCKET_MS
            } else {
                -1L
            }
        return "$title|$artist|$bucket"
    }

    internal fun normalize(value: String): String =
        value
            .trim()
            .lowercase()
            .replace(Regex("[\\p{Punct}]+"), " ")
            .replace(Regex("\\s+"), " ")
            .trim()

    /**
     * Prefer the copy that looks like the "real" library rip: more plays, lives
     * under Music/, larger file, has art. Never prefers deleting.
     */
    internal val keepComparator: Comparator<TrackEntity> =
        compareByDescending<TrackEntity> { it.playCount }
            .thenByDescending { musicFolderScore(it.folderPath) }
            .thenByDescending { it.size }
            .thenByDescending { if (it.artworkUri.isNullOrBlank()) 0 else 1 }
            .thenBy { it.id }
}

private fun musicFolderScore(folderPath: String): Int {
    val path = folderPath.lowercase()
    return when {
        path == "music" || path.startsWith("music/") -> 2
        path.contains("/music/") || path.endsWith("/music") -> 1
        else -> 0
    }
}

File: ios/Simple MP3/Data/DuplicateDetector.swift
//
//  DuplicateDetector.swift
//  Simple MP3
//

import Foundation

enum DuplicateDetector {
    static let durationBucketMs: Int64 = 2_000

    struct Group: Identifiable, Sendable {
        var id: String { key }
        var key: String
        var tracks: [Track]
        var keepId: String

        var extras: [Track] { tracks.filter { $0.id != keepId } }
        var keeper: Track { tracks.first { $0.id == keepId } ?? tracks[0] }
    }

    static func findGroups(_ tracks: [Track]) -> [Group] {
        let songs = tracks.filter { $0.source != .stream && !$0.isHidden }
        var buckets: [String: [Track]] = [:]
        for track in songs {
            guard let key = fingerprint(track) else { continue }
            buckets[key, default: []].append(track)
        }
        return buckets.values
            .filter { $0.count >= 2 }
            .map { group in
                let ordered = group.sorted(by: keepBefore)
                return Group(
                    key: fingerprint(ordered[0]) ?? ordered[0].id,
                    tracks: ordered,
                    keepId: ordered[0].id
                )
            }
            .sorted { $0.keeper.title.localizedCaseInsensitiveCompare($1.keeper.title) == .orderedAscending }
    }

    static func fingerprint(_ track: Track) -> String? {
        let title = normalize(track.title)
        let artist = normalize(track.artist)
        if title.isEmpty || title == "unknown title" { return nil }
        if artist.isEmpty || artist == "unknown artist" { return nil }
        let bucket = track.duration > 0 ? track.duration / durationBucketMs : -1
        return "\(title)|\(artist)|\(bucket)"
    }

    static func normalize(_ value: String) -> String {
        let folded = value.lowercased()
        let stripped = folded.replacingOccurrences(of: #"[\p{Punct}]+"#, with: " ", options: .regularExpression)
        return stripped.replacingOccurrences(of: #"\s+"#, with: " ", options: .regularExpression)
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }

    private static func keepBefore(_ a: Track, _ b: Track) -> Bool {
        if a.playCount != b.playCount { return a.playCount > b.playCount }
        let ascore = musicFolderScore(a.folderPath)
        let bscore = musicFolderScore(b.folderPath)
        if ascore != bscore { return ascore > bscore }
        if a.size != b.size { return a.size > b.size }
        let aArt = a.artworkUri?.isEmpty == false
        let bArt = b.artworkUri?.isEmpty == false
        if aArt != bArt { return aArt && !bArt }
        return a.id < b.id
    }

    private static func musicFolderScore(_ folderPath: String) -> Int {
        let path = folderPath.lowercased()
        if path == "music" || path.hasPrefix("music/") { return 2 }
        if path.contains("/music/") || path.hasSuffix("/music") { return 1 }
        return 0
    }
}
